File: src/go/year_2023/day_05.go
package year_2023

import (
	"fmt"
	"strconv"
	"strings"
)

type MapRange struct {
	Dst  int
	Src  int
	Size int
}

type CategoryMap struct {
	DstId  string
	Ranges []MapRange
}

type Almanac struct {
	Seeds []int
	Maps  map[string]*CategoryMap
}

// Parse seeds and maps from the given data.
func NewAlmanac(data string) *Almanac {
	data = strings.TrimSpace(data)
	lines := strings.Split(data, "\n")

	a := Almanac{}
	a.parseSeeds(&lines)
	a.parseMaps(&lines)
	return &a
}

func (a *Almanac) parseSeeds(dataLines *[]string) {
	rawSeeds := strings.Split((*dataLines)[0], " ")[1:]
	a.Seeds = make([]int, len(rawSeeds))
	for i, s := range rawSeeds {
		a.Seeds[i], _ = strconv.Atoi(s)
	}
	// fmt.Fprintf(os.Stderr, "Seeds = %#v\n", a.Seeds)
}

func (a *Almanac) parseMaps(dataLines *[]string) {
	a.Maps = make(map[string]*CategoryMap)

	var cur *CategoryMap
	for _, line := range (*dataLines)[2:] {
		if cur == nil {
			var srcId string
			cur = new(CategoryMap)
			cur.Ranges = make([]MapRange, 0)

			line = strings.Replace(line, "-", " ", 2)
			fmt.Sscanf(line, "%s to %s map:", &srcId, &cur.DstId)

			a.Maps[srcId] = cur
		} else if line != "" {
			var r = MapRange{}
			fmt.Sscanf(line, "%d %d %d", &r.Dst, &r.Src, &r.Size)

			cur.Ranges = append(cur.Ranges, r)
		} else {
			cur = nil
		}
	}
}

// Convert a value "num" from category "srcId" to category "targetId".
func (a *Almanac) Convert(srcId, targetId string, num, nAfter int) (int, int) {
	// fmt.Fprintf(os.Stderr, "%d -> ", num)
	if srcId == targetId {
		// fmt.Fprintf(os.Stderr, "%d | Skippable: %d\n", num, nAfter)
		return num, nAfter
	}

	var nBefore int
	for _, rng := range a.Maps[srcId].Ranges {
		if rng.Src <= num && num < rng.Src+rng.Size {
			nBefore = num - rng.Src
			nAfter = min(nAfter, rng.Size-nBefore)
			num = rng.Dst + nBefore
			break
		}
	}

	return a.Convert(a.Maps[srcId].DstId, targetId, num, nAfter)
}

func (a *Almanac) String() string {
	var sb = strings.Builder{}
	for src, m := range a.Maps {
		fmt.Fprintf(&sb, "CategoryMap: %s -> %s\n  Ranges:\n", src, m.DstId)
		for _, r := range m.Ranges {
			fmt.Fprintf(&sb, "    %d -> %d [%d]\n", r.Src, r.Dst, r.Size)
		}
	}
	return sb.String()
}

type Day05 struct{}

func init() {
	Solutions[05] = Day05{}
}

func (Day05) Solve1(data string) string {
	alma := NewAlmanac(data)
	// fmt.Fprintln(os.Stderr, alma.String())

	var loc, seed int
	var minLoc = 1_000_000_000_000 // Largely enough
	for _, seed = range alma.Seeds {
		loc, _ = alma.Convert("seed", "location", seed, 1)
		minLoc = min(minLoc, loc)
	}
	return fmt.Sprint(minLoc)
}

func (Day05) Solve2(data string) string {
	alma := NewAlmanac(data)
	// fmt.Fprintln(os.Stderr, alma.String())

	var loc, skip, seed, nSeeds, i int
	var minLoc = 1_000_000_000_000 // Largely enough
	for i = 0; i < len(alma.Seeds); i += 2 {
		seed = alma.Seeds[i]
		nSeeds = alma.Seeds[i+1]
		for nSeeds > 0 {
			loc, skip = alma.Convert("seed", "location", seed, nSeeds)
			minLoc = min(minLoc, loc)
			seed += skip
			nSeeds -= skip
		}
	}
	return fmt.Sprint(minLoc)
}

File: src/py/year_2023/day_05.py
import typing as t

from py.utils.debug import eprint


class MapRange(t.NamedTuple):
    dst: int
    src: int
    size: int


class Almanac:
    def __init__(self, data: str) -> None:
        """Parse seeds and maps from the given data"""
        seeds_line, _, *lines = data.strip().split("\n")
        self.seeds = [int(x) for x in seeds_line.removeprefix("seeds: ").split()]
        self.maps: dict[str, tuple[str, list[MapRange]]] = {}
        cur = None
        for line in lines:
            if cur is None:
                assert line.endswith(" map:")
                src, dst = line.removesuffix(" map:").split("-to-")  #soil-to-fertilizer map:
                cur = self.maps[src] = (dst, [])
            elif not line:
                cur = None
            else:
                cur[1].append(MapRange(*map(int, line.split())))

    def convert(self, cat: str, target: str, num: int, n_after=1) -> tuple[int, int]:
        """Convert a value "num" from category "cat" to category "target".

        n_after keeps track of the maximum amount of values that can be skipped.
        Values after "num" can be skipped because they will always be bigger.
        But map ranges are not "aligned". A value at the end of a category could
        lead to the start of another category.
        Reducing the number of values with each category is equivalent to finding
        a range that goes "straight" across all categories until target is reached.
        """
        if __debug__: eprint(num, end=" -> ")
        if cat == target:
            if __debug__: eprint(num, "skippable:", n_after)
            return num, n_after

        next_cat, ranges = self.maps[cat]
        for i_dst, i_src, size in ranges:
            if i_src <= num < i_src + size:
                n_before = num - i_src
                n_after = min(n_after, size - n_before)
                num = i_dst + n_before
                break

        return self.convert(next_cat, target, num, n_after)

    def __str__(self):
        return "".join(
            f"[Category Map: {src} -> {dst}]\n"
            + "".join(f"    {r}\n" for r in ranges)
            for src, (dst, ranges) in self.maps.items()    
        )


def solve_1(data: str):
    alma = Almanac(data)
    if __debug__: eprint(alma)

    minloc = 10**15  # Largely enough
    for seed in alma.seeds:
        loc, _ = alma.convert("seed", "location", seed)
        minloc = min(minloc, loc)
    return minloc


def solve_2(data: str):
    alma = Almanac(data)
    if __debug__: eprint(alma)

    minloc = 10**15  # Largely enough
    for i in range(0, len(alma.seeds), 2):
        seed, n_seeds = alma.seeds[i:i+2]
        while n_seeds > 0:
            loc, skip = alma.convert("seed", "location", seed, n_seeds)
            minloc = min(minloc, loc)
            seed += skip
            n_seeds -= skip
    return minloc


if __name__ == "__main__":
    import sys

    data = sys.stdin.read()
    print("\n[PART 1]", solve_1(data), sep="\n")
    print("\n[PART 2]", solve_2(data), sep="\n")
